spec/models/form_spec.rb

```diff
@@ -205,32 +205,140 @@ def answer_local_authority(lettings_log)
+          log.form.reset_not_routed_questions_and_invalid_answers(log)
+          expect(log.stairbought).to be nil
+        end
       end
+    end
+
+    context "when there are multiple free user input questions for attribute X" do
+      context "and attribute Y is changed such that a different question for X is routed to" do
+        let(:log) { FactoryBot.create(:sales_log, :saledate_today, :shared_ownership, :privacy_notice_seen, jointpur: 1, jointmore: 2, hholdcount: expected_hholdcount) }
+        let(:expected_hholdcount) { 2 }
+
+        it "the value of this attribute is not cleared" do
+          log.jointpur = 2
+          log.form.reset_not_routed_questions_and_invalid_answers(log)
+          expect(log.hholdcount).to eq expected_hholdcount
+        end
+      end
+
+      context "and attribute Y is changed such that no questions for X are routed to" do
+        let(:log) { FactoryBot.create(:sales_log, :shared_ownership_setup_complete, value: initial_value) }
+        let(:initial_value) { 200_000.to_d }
+
+        it "the value of this attribute is cleared" do
+          expect(log.value).to eq initial_value
+          log.ownershipsch = 2
+          log.form.reset_not_routed_questions_and_invalid_answers(log)
+          expect(log.value).to be nil
+        end
+      end
+    end
+
+    context "when a value is changed such that a checkbox question is no longer routed to" do
+      let(:log) { FactoryBot.create(:lettings_log, :about_completed, startdate: now, reasonpref: 1, rp_homeless: 1, rp_medwel: 1, rp_hardship: 1) }
+
+      it "all attributes relating to that checkbox question are cleared" do
+        expect(log.rp_homeless).to be 1
+        log.reasonpref = 2
+        log.form.reset_not_routed_questions_and_invalid_answers(log)
+        expect(log.rp_homeless).to be nil
+        expect(log.rp_medwel).to be nil
+        expect(log.rp_hardship).to be nil
+      end
+    end
+
+    context "when an attribute is derived, but no questions for that attribute are routed to" do
+      let(:log) { FactoryBot.create(:sales_log, :outright_sale_setup_complete, value: 200_000) }
+
+      it "the value of this attribute is not cleared" do
+        expect(log.deposit).to be nil
+        log.update!(mortgageused: 2)
+        expect(log.form.questions.any? { |q| q.id == "deposit" && q.page.routed_to?(log, nil) }).to be false
+        expect(log.deposit).not_to be nil
+      end
+    end
+
+    context "when an attribute is related to a callback question with no set answer options, and no questions for that attribute are routed to" do
+      let(:location) { FactoryBot.create(:location) }
+      let(:log) { FactoryBot.create(:lettings_log, :startdate_today) }
 
-      it "does not route to the page" do
-        expect(form.invalidated_pages(lettings_log).map(&:id)).to include("outstanding_amount_known")
+      # Pages::PropertyPostcode and questions inside have been removed from form. do we not want to delete and migration delete_column?
```

Comment:
this question is mostly directed at @bibblobcode 
On reflection I'm assuming that the columns we may want to keep for existing data/migrated data, but since the lettings forms made from `Questions::BlaBla` are only for 23/24 maybe they can be deleted?

Reply:
yes the latter. the Pages::PropertyPostcode is no longer rendered anywhere. and then therefore the Questions::Bla classes inside that